lib/SymbolGraphGen/SymbolGraphGen.cpp

```diff
@@ -22,10 +22,21 @@ using namespace symbolgraphgen;
 namespace {
 int serializeSymbolGraph(SymbolGraph &SG,
                          const SymbolGraphOptions &Options) {
-  SmallString<256> FileName(SG.M.getNameStr());
-  if (SG.ExtendedModule.hasValue()) {
-    FileName.push_back('@');
-    FileName.append(SG.ExtendedModule.getValue()->getNameStr());
+  SmallString<256> FileName;
+  if (SG.DeclaringModule.hasValue()) {
+    // Save a cross-import overlay symbol graph as `MainModule@BystandingModule[@BystandingModule...].symbols.json`
```

Comment:
Beware that one declaring module/bystanding module pair can have several overlays. For instance, although nothing in the Apple SDKs currently does this, it's perfectly valid for `MapKit.swiftcrossimport/SwiftUI.swiftoverlay` to say:

```
version: 1
modules:
  - name: _MapKit_SwiftUI
  - name: _MapKit_SwiftUI_DeviceOnly
```

And then SymbolGraphGen would try to write information for both of these modules to `MapKit@SwiftUI.symbols.json`.

This could be a problem unless:

1. Some higher-level tool will account for this and merge the same-named files together, or
2. The `getDeclaringModuleIfCrossImportOverlay()` and `getRequiredBystandersIfCrossImportOverlay()` helpers already account for this somehow. (I believe @nathawes implemented these; he would know better than I.)

If that *is* a problem, then consider adding the overlay name to the filename; `DeclaringModule@BystandingModule[@BystandingModule...]@OverlayModule.symbols.json` should never conflict with other overlays. (On the other hand, there could be duplicate files if the same overlay module is declared in several different .swiftoverlay files; that's technically valid but even weirder, and I'm not sure why anyone would do it in practice.)

Reply:
That's an excellent point! I think, given the way i approach it in this PR, there are two approaches:

1. Add the overlay module name to the symbol graph file name, like you mention.
2. Merge the cross-import overlay symbol graphs in the symbol graph tool when there are multiple overlay modules for the same sets of declaring/bystander modules.

I don't remember offhand if the symbol graph tool has a way to load an existing symbol graph, so i'm not sure how feasible this is to implement here. Given that, i think going for (1) would be the simplest option, so i'll go for that.